### DISCUSSION TITLE
xhttp过cdn时nginx出现错误代码问题

### DISCUSSION THREAD

**Discussion by wtfr-dot** [XTLS/Xray-core#4162] [Q&A]
我的网络代理是这样布局的：vless+tls+xhttp---cf的cdn---vps入站回落trojan---trojan回落nginx---nginx反代xhttp，之前采用gRPC传输层协议，工作比较稳定，为了响应大佬的倡议，将其迁移到xhttp，经过一番设置后，目前网络能正常工作，但是在nginx的访问日志里发现大量的400错误，以前用gRPC几乎没有这种情况，另外499错误也增多了，像下面这样：
172.69.34.182:60024 - - [14/Dec/2024:02:20:42 -0800]200 189 "-" "Go-http-client/2.0" 
172.69.34.182:60024 - - [14/Dec/2024:02:20:42 -0800]200 189 "-" "Go-http-client/2.0"
172.69.34.182:60024 - - [14/Dec/2024:02:20:42 -0800]400 0 "-" "Go-http-client/2.0" 
172.69.34.182:60024 - - [14/Dec/2024:02:20:42 -0800]400 0 "-" "Go-http-client/2.0" 
172.69.127.139:9026 - - [14/Dec/2024:02:20:48 -0800]400 0 "-" "Go-http-client/2.0" 
172.69.127.142:43340 - - [14/Dec/2024:02:20:48 -0800]200 0 "-" "Go-http-client/2.0"
172.69.127.148:42662 - - [14/Dec/2024:02:20:48 -0800]400 0 "-" "Go-http-client/2.0" 
172.69.127.139:9026 - - [14/Dec/2024:02:20:48 -0800]200 0 "-" "Go-http-client/2.0" 
172.69.34.182:60024 - - [14/Dec/2024:02:20:50 -0800]400 0 "-" "Go-http-client/2.0"
172.69.34.182:60024 - - [14/Dec/2024:02:20:50 -0800]200 5761 "-" "Go-http-client/2.0"
162.158.186.185:16962 - - [14/Dec/2024:02:21:03 -0800]400 0 "-" "Go-http-client/2.0"
162.158.186.184:9806 - - [14/Dec/2024:02:21:03 -0800]200 4599 "-" "Go-http-client/2.0" 
172.70.207.15:49354 - - [14/Dec/2024:02:21:09 -0800]200 1088 "-" "Go-http-client/2.0" 
172.70.207.15:49354 - - [14/Dec/2024:02:21:09 -0800]400 0 "-" "Go-http-client/2.0"
172.70.211.233:47974 - - [14/Dec/2024:02:21:13 -0800]400 0 "-" "Go-http-client/2.0"
172.70.211.233:47974 - - [14/Dec/2024:02:21:13 -0800]200 154530 "-" "Go-http-client/2.0"
162.158.186.185:16962 - - [14/Dec/2024:02:21:17 -0800]400 0 "-" "Go-http-client/2.0"
162.158.186.184:9806 - - [14/Dec/2024:02:21:17 -0800]200 7082 "-" "Go-http-client/2.0" 
172.69.34.182:60024 - - [14/Dec/2024:02:21:17 -0800]200 0 "-" "Go-http-client/2.0" 
172.69.34.182:60024 - - [14/Dec/2024:02:21:17 -0800]200 0 "-" "Go-http-client/2.0"
172.69.34.182:60024 - - [14/Dec/2024:02:21:17 -0800]200 0 "-" "Go-http-client/2.0" 
172.69.34.182:60024 - - [14/Dec/2024:02:21:17 -0800]499 0 "-" "Go-http-client/2.0"
162.158.186.184:9806 - - [14/Dec/2024:02:21:19 -0800]400 0 "-" "Go-http-client/2.0"
162.158.186.185:16962 - - [14/Dec/2024:02:21:19 -0800]200 4599 "-" "Go-http-client/2.0" 
172.69.34.182:60024 - - [14/Dec/2024:02:21:22 -0800]499 0 "-" "Go-http-client/2.0" 
172.70.207.15:49354 - - [14/Dec/2024:02:21:23 -0800]499 0 "-" "Go-http-client/2.0" 
172.70.207.15:49354 - - [14/Dec/2024:02:21:26 -0800]499 0 "-" "Go-http-client/2.0" 
162.158.186.185:16962 - - [14/Dec/2024:02:21:26 -0800]499 0 "-" "Go-http-client/2.0" 
162.158.186.184:9806 - - [14/Dec/2024:02:21:26 -0800]499 0 "-" "Go-http-client/2.0"
172.70.207.15:49354 - - [14/Dec/2024:02:21:26 -0800]499 0 "-" "Go-http-client/2.0" 
172.70.207.15:49354 - - [14/Dec/2024:02:21:26 -0800]499 0 "-" "Go-http-client/2.0" 

请大佬帮忙分析一下是哪里设置的问题，我的nginx的fallback配置如下：

server {
    listen unix:/dev/shm/h1.socket proxy_protocol default_server;
    listen unix:/dev/shm/h2c.socket proxy_protocol default_server;
    set_real_ip_from unix:;
    real_ip_header proxy_protocol;
    server_name _;
    return 400;
}

# HTTP1 UDS listener
server {
    listen unix:/dev/shm/h1.socket proxy_protocol; 
	set_real_ip_from unix:;
    real_ip_header proxy_protocol;
    server_name  xxx
	
    location / {
        # POST returns 404 when negotiation fails
        # if ($request_method != "POST") {
        #    return 404;
        # }
        client_body_buffer_size 1m;
        client_body_timeout 1h;
        client_max_body_size 0;
        proxy_pass https://xxxxxx
	}
}

# HTTP2 UDS listener
server {
    listen unix:/dev/shm/h2c.socket proxy_protocol; 
    set_real_ip_from unix:;
    real_ip_header proxy_protocol;
    server_name  xxx

    location /xxxx/ { 
  			grpc_read_timeout                 	5m;
		grpc_send_timeout                 	5m;
		grpc_set_header X-Forwarded-For   	$proxy_add_x_forwarded_for;
		grpc_set_header X-Real-IP         	$remote_addr;
		grpc_set_header X-Real-IP         	$remote_port;
		client_body_buffer_size 			0;
        client_body_timeout 				5m;
		grpc_pass unix:/dev/shm/xhttp.socket;
    }

    location /xxxxxxx { 
        # return 404 if HTTP Method is not POST
        if ($request_method != "POST") {
            return 404;
        }
		
		grpc_buffer_size                  8k;
		grpc_connect_timeout              70s;
		grpc_socket_keepalive             on;
		grpc_read_timeout                 1h;
		grpc_send_timeout                 1h;
		grpc_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;		
		grpc_set_header X-Real-IP         $remote_addr;
		grpc_set_header X-Real-IP         $remote_port;
		client_body_buffer_size           512k;
        client_body_timeout               1h;
		grpc_pass unix:/dev/shm/grpc.socket;
	}
	
    location / { 
        client_body_buffer_size 1m;
        client_body_timeout 1h;
        client_max_body_size 0;
        proxy_pass https://xxxxxx
    }	
	
}

**Comment by xqzr** [comment]
https://github.com/XTLS/Xray-core/pull/4163#issuecomment-2543193465

**Comment by junlongzzz** [comment]
Caddy 反代一样，只要使用 xhttp mode 为 stream-* 的，请求就会抛出大量类似 499 错误，虽不影响使用，~~但很碍眼~~，不过 CDN 使用直连也一样错误响应

错误日志之一：
```json
{
    "level": "error",
    "ts": 1734270468.0393834,
    "logger": "http.log.error.log0",
    "msg": "stream error: stream ID 147; CANCEL",
    "request": {
        "remote_ip": "***",
        "remote_port": "59886",
        "client_ip": "***",
        "proto": "HTTP/2.0",
        "method": "POST",
        "host": "***",
        "uri": "/***/cc858be3-0aee-447f-8db6-1c0f4120c390?x_padding=000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
        "headers": {
            "Cf-Ray": [
                "8f26e27d7d42c82f-HKG"
            ],
            "Cdn-Loop": [
                "cloudflare; loops=1"
            ],
            "X-Forwarded-Proto": [
                "https"
            ],
            "Cf-Visitor": [
                "{\"scheme\":\"https\"}"
            ],
            "Content-Type": [
                "application/grpc"
            ],
            "User-Agent": [
                "Go-http-client/2.0"
            ],
            "Accept-Encoding": [
                "identity"
            ],
            "Cf-Ipcountry": [
                "***"
            ],
            "X-Forwarded-For": [
                "***"
            ],
            "Cf-Connecting-Ip": [
                "***"
            ],
            "Te": [
                "trailers"
            ]
        },
        "tls": {
            "resumed": false,
            "version": 772,
            "cipher_suite": 4865,
            "proto": "h2",
            "server_name": "***"
        }
    },
    "duration": 29.991073754,
    "status": 502,
    "err_id": "b9iezk1si",
    "err_trace": "reverseproxy.statusError (reverseproxy.go:1269)"
}
```

**Comment by wtfr-dot** [comment]
看来不是设置问题了，谢谢各位！
